update example: e293 (changed branch)[duda] (#2)
DevExpress-Examples/aspxgridview-dataitemtemplate-runtime: Readme.md

<!-- default badges end -->
# ASP.NET - Create a DataItemTemplate for an ASPxGridView Сolumn at Runtime

This example demonstrates how to create a [column cell template](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) at runtime. Grid cells display the [ASPxHyperLink](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxHyperLink) control created in the template.

![ASPxGridView-DataItemTemplate](ASPxGridView-DataItemTemplate.png)

## Files to look at:

* [Default.aspx](./CS/HyperlinkColumn/Default.aspx) (VB: [Default.aspx](./VB/HyperlinkColumn/Default.aspx))
* [Default.aspx.cs](./CS/HyperlinkColumn/Default.aspx.cs) (VB: [Default.aspx.vb](./VB/HyperlinkColumn/Default.aspx.vb))
* [details.aspx](./CS/HyperlinkColumn/details.aspx) (VB: [details.aspx](./VB/HyperlinkColumn/details.aspx))
* [details.aspx.cs](./CS/HyperlinkColumn/details.aspx.cs) (VB: [details.aspx.vb](./VB/HyperlinkColumn/details.aspx.vb))

## Implementation Details

Declare a custom class that implements the [ITemplate](https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.itemplate) interface.

> In the class's [InstantiateIn](https://docs.microsoft.com/ru-ru/dotnet/api/system.web.ui.itemplate.instantiatein) method, create controls that make up the template and add them to the [control collection](https://docs.microsoft.com/ru-ru/dotnet/api/system.windows.forms.control.controls) of a container passed as this method's parameter. A container type is different for each template type. For a [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) object, the container is of the [GridViewDataItemTemplateContainer](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataItemTemplateContainer) type. Use the container's properties to obtain a row's information. For example, use the **KeyValue** property to get a row key (`container.KeyValue`).

Reviewer: ```suggestion
In the class's [InstantiateIn](https://docs.microsoft.com/ru-ru/dotnet/api/system.web.ui.itemplate.instantiatein) method, create controls that make up the template and add them to the [control collection](https://docs.microsoft.com/ru-ru/dotnet/api/system.windows.forms.control.controls) of a container passed as this method's parameter. The container type is different for each template type. For a [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) object, the container is of the [GridViewDataItemTemplateContainer](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataItemTemplateContainer) type. Use the container's properties to obtain a row's information. For example, use the **KeyValue** property to get a row key (`container.KeyValue`).
```

[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/E293)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->
# ASP.NET - Create a DataItemTemplate for an ASPxGridView column at Runtime

> This example demonstrates how to create a [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) at runtime.

Reviewer 2: Может стоит сказать, что реализует создаваемый DataItemTemplate?

Author: Т.е. что-то вроде этого?

`This example demonstrates how to create a [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) at runtime. The template of this type allows you to customize data cells in grid-like controls.`

Reviewer 2: Не совсем то. Я как пользователь хотел бы знать что из себя представляет создаваемый темплейт. Что за ASPxHyperLink там используется.  И т.п. кратенько.

![ASPxGridView-DataItemTemplate](ASPxGridView-DataItemTemplate.png)

<!-- default file list -->
## Files to look at:

* [Default.aspx](./CS/HyperlinkColumn/Default.aspx) (VB: [Default.aspx](./VB/HyperlinkColumn/Default.aspx))
* [Default.aspx.cs](./CS/HyperlinkColumn/Default.aspx.cs) (VB: [Default.aspx.vb](./VB/HyperlinkColumn/Default.aspx.vb))
* [details.aspx](./CS/HyperlinkColumn/details.aspx) (VB: [details.aspx](./VB/HyperlinkColumn/details.aspx))
* [details.aspx.cs](./CS/HyperlinkColumn/details.aspx.cs) (VB: [details.aspx.vb](./VB/HyperlinkColumn/details.aspx.vb))
<!-- default file list end -->

## Implementation Details

> Use a column's [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) property to specify a template for the column's cells. This property gets an object with an [ITemplate](https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.itemplate) standard interface. To create this object, declare a custom class and implement the [ITemplate](https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.itemplate) interface in it.

Reviewer 2: В целом этот абзац вполне корректен. Цепляет второе предложение, где слово `with` не однозначно. Предложение может означать, что 'object' возвращается **вместе** с неким интерфейсом 'ITemplate'.  

В третьем предложении можно избавиться от 'in it', сказав: declare a custom class **that** implements...

Первое предложение в этом абзаце является неким повторением последнего абзаца ("To specify a template for a column's cells,..." Подумай, как от этого можно избавиться.

Author: 1. По поводу with: я использовал его, чтобы уйти от повтора implement. Иными словами "принимает объект, в котором реализован такой-то интерфейс"
Можем ли мы в таком случае использовать `that has`? Т.е. "объект, который имеет такой-то интерфейс"

2. Хорошо, поправлю

3. Во втором случае я это использовал больше как этап
Мне кажется можно попробовать отойти от конкретной проперти, а также не разделять абзац, потому что предложение достаточно маленькое:
`Use the container's properties to obtain a row's information. For example, use the **KeyValue** property to get a row key (container.KeyValue). An instance of the created class can serve as a template for a column's сells.`
Как думаешь?

Reviewer 2: > я это использовал больше как этап

Всё верно. Последний абзац - это этап. Поэтому и надо сделать так, чтобы другое предложение (самое первое предложение после `Implementation Details`) не выглядело как этап.
Самое простое - это оформить шаги списком.

> In the class's [InstantiateIn](https://docs.microsoft.com/ru-ru/dotnet/api/system.web.ui.itemplate.instantiatein) method, create controls and add them to the [control collection](https://docs.microsoft.com/ru-ru/dotnet/api/system.windows.forms.control.controls) of a container passed to this method as the **Control** parameter. A container type is different for each template type. For the [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate), the parent container is the [GridViewDataItemTemplateContainer](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataItemTemplateContainer). Use the container's properties to obtain a row's information. For example, use the **KeyValue** property to get a row key (`container.KeyValue`).

Reviewer 2: > "create controls"
какие контролы надо создать?

> "passed to this method as the **Control** parameter."
можно упростить:  passed as this method's **Control** parameter.

> "the parent container is the [GridViewDataItemTemplateContainer]."
Нужно тут использовать слово 'parent' ? Оно больше нигде здесь не упоминается...
После GridViewDataItemTemplateContainer нет существительного, обозначающего про что ты говоришь (class, object,...). Если ты имеешь в виду 'object' - то уместен тогда неопределенный артикль.

Author: 1. В зависимости от нужд пользователя, он задает какие-то контролы для ячейки, изначально было create required controls, но тут required особо не дает какой-то прям полезной дополнительной информации, как мне кажется

2. Отлично, спасибо

3. Фактически любой контейнер, связанный с темплейтом, - это его парент контейнер, так что я согласен, можно опустить parent. 
По поводу артикля я понял, хорошо, поправлю 
Я рассматривал этот момент со стороны того, что у нас есть теплейт конкретного типа, и для него есть контейнер конкретного типа

> To specify a template for a column's cells, create an instance of a template class and assign it to the column's [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) property.

Reviewer 2: см. комментарий к первому абзацу

Reviewer 2: Здесь можно убрать часть предложения "To specify a template for a column's cells". Смысл не потеряется.

[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->
# ASP.NET - Create a DataItemTemplate for an ASPxGridView Сolumn at Runtime

This example demonstrates how to create a [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) at runtime. Grid cells display the [ASPxHyperLink](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxHyperLink) control created in this template.

![ASPxGridView-DataItemTemplate](ASPxGridView-DataItemTemplate.png)

<!-- default file list -->
## Files to look at:

* [Default.aspx](./CS/HyperlinkColumn/Default.aspx) (VB: [Default.aspx](./VB/HyperlinkColumn/Default.aspx))
* [Default.aspx.cs](./CS/HyperlinkColumn/Default.aspx.cs) (VB: [Default.aspx.vb](./VB/HyperlinkColumn/Default.aspx.vb))
* [details.aspx](./CS/HyperlinkColumn/details.aspx) (VB: [details.aspx](./VB/HyperlinkColumn/details.aspx))
* [details.aspx.cs](./CS/HyperlinkColumn/details.aspx.cs) (VB: [details.aspx.vb](./VB/HyperlinkColumn/details.aspx.vb))
<!-- default file list end -->

## Implementation Details

> A column's [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) property allows you to specify a template for the column's cells as follows.

Reviewer: `as follows` As follows what? Is there a list?

> A column's [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) property allows you to specify a template for the column's cells as follows.

Author: ```suggestion
A column's [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate) property allows you to specify a template for the column's cells as described below.
```

The property gets an object that has an [ITemplate](https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.itemplate) standard interface. To create this object, declare a custom class and implement the [ITemplate](https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.itemplate) interface in it.

> In the class's [InstantiateIn](https://docs.microsoft.com/ru-ru/dotnet/api/system.web.ui.itemplate.instantiatein) method, create controls that will make up the template and add them to the [control collection](https://docs.microsoft.com/ru-ru/dotnet/api/system.windows.forms.control.controls) of a container passed as this method's parameter. A container type is different for each template type. For the [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate), the container is a [GridViewDataItemTemplateContainer](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataItemTemplateContainer). Use the container's properties to obtain a row's information. For example, use the **KeyValue** property to get a row key (`container.KeyValue`).

Reviewer: ```suggestion
In the class's [InstantiateIn](https://docs.microsoft.com/ru-ru/dotnet/api/system.web.ui.itemplate.instantiatein) method, create controls that make up the template and add them to the [control collection](https://docs.microsoft.com/ru-ru/dotnet/api/system.windows.forms.control.controls) of a container passed as this method's parameter. A container type is different for each template type. For the [DataItemTemplate](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataColumn.DataItemTemplate), the container is a [GridViewDataItemTemplateContainer](https://docs.devexpress.com/AspNet/DevExpress.Web.GridViewDataItemTemplateContainer). Use the container's properties to obtain a row's information. For example, use the **KeyValue** property to get a row key (`container.KeyValue`).
```
